Issue #21: Implementation of local-as command

lyubov-popova opened the pull request:
Change-Id: Iad717eab3f16a1c722d79f55bec0008bf0529e2c

https://github.com/MERAprojects/ops-build/issues/21

Following BGP commands implemented in commit:
neighbor x.x.x.x local-as <1-4294967295>
no neighbor x.x.x.x local-as <1-4294967295>

kontorschikov commented on vtysh/bgp_vty.c:
```diff
@@ -3228,6 +3249,103 @@ DEFUN(no_neighbor_peer_group_remote_as,
+    if (!ovs_bgp_neighbor) {
+        ABORT_DB_TXN(txn, BGP_ERR_NEIGHBOR_IS_NOT_CONFIGURED);
+    }
+
```
Missing check for peer-group membership?
` if(ovs_bgp_neighbor->bgp_peer_group) {`
 `ABORT_DB_TXN(txn, BGP_ERR_INVALID_FOR_PEER_GROUP_MEMBER);}`

kontorschikov commented on vtysh/bgp_vty.c:
```diff
@@ -3228,6 +3249,103 @@ DEFUN(no_neighbor_peer_group_remote_as,
+    if (!ovs_bgp_neighbor) {
+        ABORT_DB_TXN(txn, BGP_ERR_NEIGHBOR_IS_NOT_CONFIGURED);
+    }
+
```
Missing check for peer-group membership for "no neighbor local-as" command?
` if(ovs_bgp_neighbor->bgp_peer_group) {`
 `ABORT_DB_TXN(txn, BGP_ERR_INVALID_FOR_PEER_GROUP_MEMBER);}`

lyubov-popova replied:
yes, fixed

mfwlarionov commented on vtysh/bgp_vty.c:
```diff
@@ -2708,6 +2708,30 @@ is_ibgp_peer(const struct ovsrec_bgp_neighbor *ovs_bgp_neighbor,
+void
+remove_parameters_invalid_for_ibgp_peer(const struct ovsrec_bgp_neighbor *ovs_bgp_neighbor)
+{
+    if (ovs_bgp_neighbor->remove_private_as) {
```
I suggest instead of operation with a pointer used with a counter operation.  
In this case "ovs_bgp_neighbor->remove_private_as" replace  "ovs_bgp_neighbor->n_remove_private_as"

mfwlarionov commented on vtysh/bgp_vty.c:
```diff
@@ -2708,6 +2708,30 @@ is_ibgp_peer(const struct ovsrec_bgp_neighbor *ovs_bgp_neighbor,
+        ovsrec_bgp_neighbor_set_remove_private_as(ovs_bgp_neighbor, NULL, 0);
+    }
+
+    if (ovs_bgp_neighbor->local_as) {
```
Similarly suggest instead of operation with a pointer used with a counter operation.,

mfwlarionov commented on vtysh/bgp_vty.c:
```diff
@@ -3228,6 +3249,103 @@ DEFUN(no_neighbor_peer_group_remote_as,
+        ABORT_DB_TXN(txn, BGP_ERR_LOCAL_AS_ALLOWED_ONLY_FOR_EBGP);
+    }
+
+    if (local_as == asn) {
```
Suggest аdd analysis of the impossibility of assigning "local_as" values are equal "remote as" for neighbor. If the neighbor is a group that respectively  need analysis for all members of the group.

lyubov-popova replied:
Added below for peer-group (BGP_ERR_CANNOT_HAVE_LOCAL_AS_SAME_REMOTE_AS) and all neighbors

mfwlarionov commented on vtysh/bgp_vty.c:
```diff
@@ -3520,6 +3636,10 @@ cli_neighbor_set_peer_group_cmd_execute(char *vrf_name, const char *ip_addr,
 	                               (int64_t *)&tim_val, ovs_bgp_peer_group->n_timers);
     }
 
+    if (ovs_bgp_peer_group->local_as) {
```
Similarly suggest instead of operation with a pointer used with a counter operation

mfwlarionov commented on vtysh/vtysh_ovsdb_router_context.c:
```diff
@@ -195,13 +195,23 @@ static void vtysh_router_context_bgp_print_nbr_cfg(vtysh_ovsdb_cbmsg_ptr p_msg,
         }
     }
+
+    if (nbr_table->local_as) {
```
Similarly suggest instead of operation with a pointer used with a counter operation

AKochin replied:
Do you suggest to do it for all the cases in this function or just local-as?

To keep the common style we the proper way is to replace all such checks in this file(or at least in this function) or let it be as it is now. I see no major point to apply change that you've suggested.

mfwlarionov replied:
Yes, I propose to maintain this style everywhere. In my opinion the number of records is more adequate criterion, than the value of the pointer.

lyubov-popova commented:
Hi!
Comments are applied and merged into 1 new commit 411200e:

1) _Use number of records instead of the value of the pointer_ - Fixed for local_as

2) _Add analysis of the impossibility of assigning "local_as" values are equal "remote as" for neighbor_ - Fixed for peer-group and neighbors in `cli_neighbor_local_as_cmd_execute`. Similar check added to `cli_neighbor_set_peer_group_cmd_execute` before setting of local_as for peer-group member. Need if remote-as is configured for peer-group member, not for peer-group.

3) Added unset of local_as for peer-group member in `cli_neighbor_set_peer_group_cmd_execute` - this parameter is inherited from the peer-group (e.g., in Cisco). 

4) Added unset of local_as in `cli_no_neighbor_set_peer_group_cmd_execute`

lyubov-popova commented:
Hi!
Comments are applied in 0dd382f

lyubov-popova commented:
Replaced local_as == *ovs_bgp_neighbor->remote_as to compare_neighbor_remote_as_with_asn(ovs_bgp_neighbor, local_as) in e53cc58
